### PR TITLE
[FileSystem] deprecate use of READ_CHUNKED flag

### PR DESCRIPTION
**Pull request by thexai**
## Description
[FileSystem] deprecate use of READ_CHUNKED flag

This is a follow up of https://github.com/xbmc/xbmc/pull/25128 or better alternative as fixes the same in more elaborated way. Also fixes other potential issues or "unwanted behavior changes" due the main refactor in StreamBuffer logic.

## Motivation and context
As commented in the other PR, `READ_AUDIO_VIDEO` flag fits better into `CFile::ShouldUseStreamBuffer` method because same as FileCache, StreamBuffer only should be used in audio/video files.

Once replaced here results that `READ_CHUNKED` flag is a dead code because only is set but never read/used. Then can be deprecated.

The others changes are only small adjustments to maintain the same functionality and some potential fix.

## How has this been tested?
Runtime Windows x64 and Shield

Tested that use of FileCache and StreamBuffer is maintained same as before for most habitual scenarios: tested MKV's local/network SMB, NFS. Tested Blu-Ray BDMV folders and ISOs SMB and NFS. Adequate chunk size is used same as before: 128K on network and 6144 bytes locally. Tested DVD, DVD ISOs. 

Fix for external SRT is preserved without needing to use the `READ_NO_BUFFER` flag since when opening subtitles it does not have the `READ_AUDIO_VIDEO` flag and the file size >200 MB condition is not met either.

## What is the effect on users?
Nothing directly but prevents other potentially cases of StreamBuffer used on file contents that are not suitable (non audio/video files).

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp [hunk 1]
```diff
@@ -151,7 +151,7 @@ bool CDVDInputStreamNavigator::Open()
   {
     // if dvd image file (ISO or alike) open using libdvdnav stream callback functions
     m_pstream = std::make_unique<CDVDInputStreamFile>(
-        m_item, XFILE::READ_TRUNCATED | XFILE::READ_BITRATE | XFILE::READ_CHUNKED);
+        m_item, XFILE::READ_TRUNCATED | XFILE::READ_BITRATE | XFILE::READ_NO_CACHE);
```
Comment (arnova) [inline xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp:154]: Why is this changed to no-cache?

File: xbmc/filesystem/File.cpp [hunk 2]
```diff
@@ -66,7 +66,7 @@ bool CFile::Copy(const CURL& url2, const CURL& dest, XFILE::IFileCallback* pCall
   CURL url(url2);
   if (StringUtils::StartsWith(url.Get(), "zip://") || URIUtils::IsInAPK(url.Get()))
     url.SetOptions("?cache=no");
-  if (file.Open(url.Get(), READ_TRUNCATED | READ_CHUNKED))
+  if (file.Open(url.Get(), READ_TRUNCATED | READ_NO_CACHE | READ_NO_BUFFER))
```
Comment (arnova) [inline xbmc/filesystem/File.cpp:69]: Same: why is this changed to no-cache?

File: xbmc/filesystem/IFileTypes.h [hunk 3]
```diff
@@ -14,39 +14,44 @@ namespace XFILE
...
 
-/* indicate that that caller support read in the minimum defined chunk size, this disables internal cache then */
-  static const unsigned int READ_CHUNKED = 0x02;
+/*
+ * Indicate that caller support read in the minimum defined chunk size, this disables internal cache then.
+ * This flag is deprecated, instead use READ_NO_CACHE to disable FileCache and READ_NO_BUFFER to disable StreamBuffer
+ * On the contrary to explicitly indicate that the file has audio/video content (suitable for caching), use
+ * the READ_AUDIO_VIDEO flag.
+ */
+static const unsigned int READ_CHUNKED = 0x02;
```
Comment (arnova) [inline xbmc/filesystem/IFileTypes.h:25]: I'd either remove it all together or at least comment it out.

File: xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp (hunk 1 above)
Comment (thexai) [reply xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp:154]: If I'm not mistaken FileCache should not be used for ISO images because it produces long delays when trying to access to non-contiguous areas (small files inside the main ISO container). Then seems the original intention of `READ_CHUNKED` flag here is disable caching probably because documentation says this:

https://github.com/xbmc/xbmc/blob/adad147d4f521c3d24ce5d2325efc9e9a92f7468/xbmc/filesystem/IFileTypes.h#L19-L20

Although currently FileCache is not disabled thanks to this but rather due to not setting READ_AUDIO_VIDEO flag.

Then `READ_NO_CACHE` is not mandatory here but it doesn't seem to harm.

File: xbmc/filesystem/File.cpp (hunk 2 above)
Comment (thexai) [reply xbmc/filesystem/File.cpp:69]: Seems this method is only used to handle Zip files and in any case is not related to playback of audio/video files or Video Player in general. Then this makes sure is not used FileCache or StremBuffer to generic copy files.

File: xbmc/filesystem/IFileTypes.h (hunk 3 above)
Comment (thexai) [reply xbmc/filesystem/IFileTypes.h:25]: Yes, I agree, maybe this flag should be completely removed and update `READ_AUDIO_VIDEO` flag comment.

File: xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp (hunk 1 above)
Comment (arnova) [reply xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp:154]: Perhaps this should be in a separate commit then explaining this.

**Comment by arnova** [comment]
@thexai : 2 remarks, perhaps you're already aware but I like to mention them to provide some inside about some of the design decisions years ago:
- The READ_AUDIO_VIDEO-flag, from what I recall, was introduced years ago when double-caching was implemented. This allowed us on CFile-level to decide whether double-caching was required or not;
- Keep in mind that filecaching is primarily used to provide continuous playback of video and audio files. But it's also to improve performance when files are processed which require a lot of small backward/forward seeks. I'm not sure whether this is the case for eg. iso-files, but I assume you do ;-)

File: xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp (hunk 1 above)
Comment (thexai) [reply xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp]: OK. Removed READ_NO_CACHE flag for now. This only affects DVD ISOs, Blu-Ray ISOs are handled by libbluray instead and is not using FileCache anyway.

**Comment by thexai** [comment]
> The READ_AUDIO_VIDEO-flag, from what I recall, was introduced years ago when double-caching was implemented. This allowed us on CFile-level to decide whether double-caching was required or not;

Currently seems other flag is used for this:

https://github.com/xbmc/xbmc/blob/436befa5e98a39bcf2ed69132717e1961bb77ee2/xbmc/filesystem/FileCache.cpp#L162-L167

File: xbmc/filesystem/IFileTypes.h (hunk 3 above)
Comment (thexai) [reply xbmc/filesystem/IFileTypes.h]: Removed and updated the description of READ_AUDIO_VIDEO flag.

**Comment by arnova** [comment]
> > The READ_AUDIO_VIDEO-flag, from what I recall, was introduced years ago when double-caching was implemented. This allowed us on CFile-level to decide whether double-caching was required or not;
> 
> Currently seems other flag is used for this:
> 
> https://github.com/xbmc/xbmc/blob/436befa5e98a39bcf2ed69132717e1961bb77ee2/xbmc/filesystem/FileCache.cpp#L162-L167

Ah yes, my mistake, you're right. I confused it with the READ_MULTI_STREAM-flag.

File: xbmc/filesystem/File.cpp (hunk 2 above)
Comment (arnova) [reply xbmc/filesystem/File.cpp:69]: I guess the same thing applies here as for .iso-files: We need to consider potential small backward/forward seeks.

Comment (thexai) [reply xbmc/filesystem/File.cpp]: Removed READ_NO_CACHE flag then. Anyway seems FileCache is not used here but I don't know in what cases this code is really used: `CFile::Copy`

Comment (thexai) [reply xbmc/filesystem/File.cpp]: Seems self contained method that uses it's own buffer to copy files:

https://github.com/xbmc/xbmc/blob/426d10df804198c3402f231f72bbdf98029d61d9/xbmc/filesystem/File.cpp#L119-L187

Comment (arnova) [reply xbmc/filesystem/File.cpp]: Ah yes, that would mean that caching doesn't do much here.

**Comment by phunkyfish** [comment]
If deprecating READ_CHUNKED do we know which binary add-ons require updating? Or is that still to be figured out?

**Comment by CastagnaIT** [comment]
for example inputstream adaptive use it
https://github.com/xbmc/inputstream.adaptive/blob/4c24b00bfa6dfa4b349ae545939be594de5bb547/src/utils/CurlUtils.cpp#L225C46-L225C64
the important thing is that dont broke the curl reading in chunks
https://github.com/xbmc/inputstream.adaptive/blob/4c24b00bfa6dfa4b349ae545939be594de5bb547/src/utils/CurlUtils.cpp#L271-L276
https://github.com/xbmc/inputstream.adaptive/blob/4c24b00bfa6dfa4b349ae545939be594de5bb547/src/utils/CurlUtils.cpp#L288-L290

**Comment by thexai** [comment]
I'm not familiar with `inputstream.adaptive` code but seems that code already differentiates audio/video files from other types (maybe text/xml, http headers, etc.) with flag `ADDON_READ_AUDIO_VIDEO`:

https://github.com/xbmc/inputstream.adaptive/blob/9d7d90a4e137a9d8bade808708ef197cb8b5cdfb/src/utils/CurlUtils.cpp#L226-L227

And this is good sign.

All big files always are read with "chunks" and this is the reason flag READ_CHUNKED is confusing. The question is not if file is read with chunks (what always happens) the question if the caller is compatible with CFileStreamBuffer class in core code or if this class is compatible with type of file read. 

For this reason everything becomes easier by changing the nomenclature to `ADDON_READ_NO_BUFFER` in case of type of file read or caller is NOT compatible with `CFileStreamBuffer`.

In case of `inputstream.adaptive` seems clear: 

- If file read is audio or video sets `ADDON_READ_AUDIO_VIDEO` flag and CFileStreamBuffer is used and works. 
- If file is any other type **NOT** sets `ADDON_READ_AUDIO_VIDEO` flag and CFileStreamBuffer is NOT used (e.g: downloads a "small" text file or downloads a binary file with chunks larger than CFileStreamBuffer can handle).

However, one thing I don't like is that it seems that this flag is also set for subtitles (at least in the documentation):

https://github.com/xbmc/inputstream.adaptive/blob/9d7d90a4e137a9d8bade808708ef197cb8b5cdfb/src/utils/CurlUtils.h#L54-L59

This is something that should be reviewed because it contradicts the use of the `READ_AUDIO_VIDEO` flag in core code:

https://github.com/xbmc/xbmc/blob/f1e74e54b6284048cc906ccf632f4c81420c4bcd/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamFile.cpp#L48-L52

Except for the subtitles thing (which should work fine if there are no known bugs) I don't think any modifications are needed in `inputstream.adaptive`. Just remove the `ADDON_READ_CHUNKED` flag, which for now would remain as dead code.

**Comment by CastagnaIT** [comment]
> However, one thing I don't like is that it seems that this flag is also set for subtitles (at least in the documentation):
> https://github.com/xbmc/inputstream.adaptive/blob/9d7d90a4e137a9d8bade808708ef197cb8b5cdfb/src/utils/CurlUtils.h#L54-L59
> This is something that should be reviewed because it contradicts the use of the READ_AUDIO_VIDEO flag in core code:

on ISA there are just two distinctions, streams (that include a/v/subs) and all others kind of files (e.g. manifests, license, etc...)
in the jungle of streaming binary formats there might be a single big file where it is downloaded in chucks for only range of bytes required

keeping READ_AUDIO_VIDEO for subs what worse performance could there be? anyway downloads are done on different threads even if it results in a small slowdown it will not affect the playback

make changes for this small distinction is currently not worth it, it touches parts of the code that need to be reworked it would complicate things for future reworks, at most add a new “todo” there are already hundreds of them and no time to make reworks to fix them all

however good to know that no particular changes are required

**Comment by thexai** [comment]
I agree, it will most likely work fine and isn't worth touching... but it should be tested with this PR before merge. 

If there are any issues with subtitles we would already know the possible cause (the issues would be partial downloads or unexpected EOF, not performance issues). If we don't want to touch READ_AUDIO_VIDEO it we could also add READ_NO_BUFFER only for subtitles or even for everything (after all current code disables FileCache and sets READ_CHUNKED flag for ALL anyway).

But... if StreamBuffer is disabled for all (via READ_NO_BUFFER), not makes sense set READ_AUDIO_VIDEO flag since it would have no effect.

**Comment by CastagnaIT** [comment]
i made the changes on https://github.com/xbmc/inputstream.adaptive/pull/1721
some tests under Windows i dont see problems also subtitles are ok
i will test also on android soon just to make sure

**Comment by thexai** [comment]
> i made the changes on https://github.com/xbmc/inputstream.adaptive/pull/1721

I think the changes are fine because FileCache and StreamBuffer were not used originally. The fact that StreamBuffer could now be used is unintentional/accidental, so the safest and easiest path is disable it.

StreamBuffer would also redundant anyway because an internal buffer is already used to download with the appropriate chunk size.

File: xbmc/addons/interfaces/Filesystem.cpp [hunk 4]
```diff
@@ -137,8 +137,6 @@ unsigned int Interface_Filesystem::TranslateFileReadBitsToKodi(unsigned int addo
 
   if (addonFlags & ADDON_READ_TRUNCATED)
     kodiFlags |= READ_TRUNCATED;
-  if (addonFlags & ADDON_READ_CHUNKED)
-    kodiFlags |= READ_CHUNKED;
```
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/addons/interfaces/Filesystem.cpp:141]: Ah, nevermind, the flag is still part of the addon API, right? But with this PR setting the flag will no longre have an effect? If so, I would a a log WARNING if addons still set the deprecated flag.

[the remark this answers is not in the document]
Comment (thexai) [reply xbmc/addons/interfaces/Filesystem.cpp:141]: This PR is mainly for deprecate use in core code and fix some remaining issues.

e.g:
`CFileStreamBuffer` is used at load GUI sounds, which are local resources (not streams). Even this not cause real issues...

Add-ons API change happened in https://github.com/xbmc/xbmc/pull/25924 including updated documentation.

>  I would a a log WARNING if addons still set the deprecated flag.

This seems very good idea, I take a look...
